util: introduce `util.types.is[…]` type checks

addaleax opened the pull request:
Provide public APIs for native typechecking that is actually useful. The motivation for this is providing alternatives to userland modules that would currently rely on `process.binding('util')`.

This is based off the list in `v8.h`, with things that have better JS equivalents removed; I’m happy to bikeshed this list.

Also, for context, previous efforts:

- @jasnell’s https://github.com/jasnell/proposal-istypes
- @evanlucas’s https://github.com/nodejs/node-eps/pull/44

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

##### Affected core subsystem(s)

util

evanlucas commented on doc/api/util.md:
````diff
@@ -894,6 +894,513 @@ encoded bytes.
+For example:
+
+```js
+util.types.isAsyncFunction(function foo() {});  // Returns false
````
should this be isGeneratorFunction

evanlucas commented on doc/api/util.md:
````diff
@@ -894,6 +894,513 @@ encoded bytes.
+For example:
+
+```js
+util.types.isPromise(new Map());  // Returns true
````
isMap()

ChALkeR commented:
@evanlucas Introducing a new module is a semver-major, as it would break [npmjs.com/types](https://www.npmjs.com/package/types) module.

ChALkeR commented on lib/internal/bootstrap_node.js:
```diff
@@ -108,6 +108,28 @@
+                    'Accessing native typechecking bindings of Node ' +
+                    'directly is deprecated. ' +
+                    `Please use \`util.types.${name}\` instead.`,
+                    'DEP0XXX') :
```
If a new deprecation is introduced, it should be recorded in [doc/api/deprecations.md](https://github.com/nodejs/node/blob/master/doc/api/deprecations.md).

TimothyGu commented on src/node_types.cc:
```diff
@@ -0,0 +1,69 @@
+  V(WeakMap)                                                                  \
+  V(WeakSet)                                                                  \
+  V(ArrayBuffer)                                                              \
+  V(TypedArray)                                                               \
```
We no longer need this with `isTypedArray` being present in `lib/internal/types.js`.

TimothyGu commented on src/node_types.cc:
```diff
@@ -0,0 +1,69 @@
+namespace {
+
+#define VALUE_METHOD_MAP(V)                                                   \
+  V(External)                                                                 \
```
Four-space indentation. Ditto elsewhere.

addaleax replied:
@TimothyGu All other macro lists I can find use 2 spaces for indentation?

TimothyGu replied:
Heh. Never mind then, I guess.

addaleax commented:
@evanlucas Like @ChALkeR said, I couldn’t find a suitable name, that’s all. I don’t think `util` is a terrible home for this, though.

@evanlucas @ChALkeR @TimothyGu Addressed your nits :)

vsemozhetbyt commented on doc/api/util.md:
```diff
@@ -943,11 +1450,11 @@ added: v0.6.0
 -->
 
-> Stability: 0 - Deprecated
+> Stability: 0 - Deprecated: Use [`Array.isArray`][] instead.
```
Maybe `Array.isArray()` as per our style guide note about parentheses after methods (+ below and in the bottom reference).

addaleax replied:
@vsemozhetbyt Yup, done!

vsemozhetbyt commented on doc/api/util.md:
```diff
@@ -1014,7 +1521,7 @@ added: v0.6.0
 -->
 
-> Stability: 0 - Deprecated
+> Stability: 0 - Deprecated: Use [`util.types.isDate()`][] instead.
```
`util.types.isDate()` seems undocumented.

addaleax replied:
@vsemozhetbyt Oops, thanks for catching this!

TimothyGu commented on doc/api/util.md:
````diff
@@ -894,6 +894,526 @@ encoded bytes.
+For example:
+
+```js
+util.types.isNativeError(new Error());  // Returns true
````
Would be good to have `TypeError` (and/or some other built-in `Error`s) here too.

TimothyGu commented on doc/api/util.md:
````diff
@@ -894,6 +894,526 @@ encoded bytes.
+For example:
+
+```js
+util.types.isMapIterator((new Map())[Symbol.iterator]());  // Returns true
````
Something that may be surprising to JavaScript developers unfamiliar with iterators is that the iterators returned from `.keys()`, `.values()`, `.entries()`, and `[@@iterator]()` in fact look the same from the outside. Maybe have them here as well would aid comprehension? (Ditto for `Set`.)

TimothyGu commented on doc/api/util.md:
````diff
@@ -894,6 +894,526 @@ encoded bytes.
+util.types.isUint8ClampedArray(new Float64Array());  // Returns false
+```
+
+### util.types.isUint16Array(value)
````
Alphabetically (what other API docs do), Uint16 would sort earlier than Uint8. This is certainly as good of a place for exception as any, but I'd like to keep it consistent for now.

vsemozhetbyt replied:
The same for `util.types.isInt8Array(value)`.

addaleax replied:
I do think this sorting order is a little less weird, tbh … 😄

TimothyGu commented on doc/api/util.md:
````diff
@@ -894,6 +894,526 @@ encoded bytes.
+
+```js
+async function checkModule(wasmBuffer) {
+  const { module } = await WebAssembly.compile(wasmBuffer);
````
Would `new WebAssembly.Module(wasmBuffer)` be a better example because of its synchronicity?

If this asynchronous example is to be kept, then it seems <sup>[MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/compile#Examples) [spec](https://webassembly.github.io/spec/js-api/index.html#ref-for-dom-webassembly-compile)</sup> it should be `const module =` rather than `const { module } =`.

addaleax replied:
> If this asynchronous example is to be kept, then it seems <sup>[MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/compile#Examples) [spec](https://webassembly.github.io/spec/js-api/index.html#ref-for-dom-webassembly-compile)</sup> it should be `const module =` rather than `const { module } =`.

I don’t know why, but I don’t think V8 and MDN agree here.

TimothyGu replied:
Seems like you used `compile` here but `instantiate` in the test. Instantiate does indeed return `{ module, instance }`.

vsemozhetbyt commented on doc/api/util.md:
```diff
@@ -1176,7 +1697,8 @@ added: v0.11.5
 
-> Stability: 0 - Deprecated
+> Stability: 0 - Deprecated: Use
+> `typeof value === 'object' && typeof value !== 'function' ` instead.
```
`&& value !== null`?

vsemozhetbyt replied:
+ Is not `typeof value` can be either`'object'` or `'function'`?
```js> const v = ()=>{}
undefined
> typeof v === 'object'
false
> typeof v
'function'
```

TimothyGu commented on doc/api/util.md:
```diff
@@ -1176,7 +1697,8 @@ added: v0.11.5
 
-> Stability: 0 - Deprecated
+> Stability: 0 - Deprecated: Use
+> `typeof value === 'object' && typeof value !== 'function' ` instead.
```
This function is implemented as `value !== null && typeof value === 'object'`.

Maybe also mention that even though [functions are also objects](https://timothygu.me/es-howto/#function-object), `isObject` historically returned `false` even for functions, because it's broken.

The doc as-is doesn't make much sense.

TimothyGu commented on doc/api/util.md:
```diff
@@ -1202,7 +1724,8 @@ added: v0.11.5
 
-> Stability: 0 - Deprecated
+> Stability: 0 - Deprecated:
+> Use `typeof value !== 'object' && typeof value !== 'function'` instead.
```
The function is implemented as `value === null || typeof value !== 'object' && typeof value !== 'function'`.

vsemozhetbyt replied:
`|| value === null` ?

TimothyGu commented on doc/api/util.md:
```diff
@@ -894,6 +894,526 @@ encoded bytes.
+[`SharedArrayBuffer`][] instance.
+
+See also [`util.types.isSharedArrayBuffer()`][] and
+[`util.types.isArrayBuffer()`][].
```
nit: flip these.

TimothyGu commented on doc/api/util.md:
```diff
@@ -894,6 +894,526 @@ encoded bytes.
+added: REPLACEME
+-->
+
+Returns `true` if the value is a string object, e.g. created
```
s/string/symbol/

Also, no need for `e.g.` as `Object()` is the only way to get symbol objects in JavaScript (`new Symbol()` always throws an error).

TimothyGu commented on test/parallel/test-util-types.js:
```diff
@@ -0,0 +1,106 @@
+
+(async function() {
+  const buffer = fixtures.readSync('test.wasm');
+  const { module } = await WebAssembly.instantiate(buffer, {});
```
Ditto as the documentation change. You could use `WebAssembly.Module`'s constructor to avoid the async-ness. If not, we don't really need to instantiate it. `WebAssembly.compile` should be sufficient.

addaleax replied:
Yeah, I simply didn’t know you could just, you know, use `new` to construct a Module. ;)

TimothyGu commented on test/parallel/test-util-types.js:
```diff
@@ -0,0 +1,106 @@
+const exclude = [
+  'Undefined', 'Null', 'NullOrUndefined', 'True', 'False', 'Name', 'String',
+  'Symbol', 'Function', 'Array', 'Object', 'Boolean', 'Number', 'Int32',
+  'Uint32', 'GeneratorObject'
```
Do we have an existing check for Generator objects? In any case, I am interested in seeing a reliable JS equivalent for `isGeneratorObject`.

addaleax replied:
@TimothyGu I’d guess you’d usually just check for presence of the methods? It’s not “reliable” in the sense that the V8 method is, but I didn’t see much benefit in including this, plus (and I could be completely wrong about that) I’d guess that this doesn’t work on transpiled generator functions anyway?

I’m happy to include it if you think it makes sense.

TimothyGu replied:
> I’d guess you’d usually just check for presence of the methods?

Doesn't this kind of defeats the whole point of having these functions available?

> I’d guess that this doesn’t work on transpiled generator functions anyway?

True, but this wouldn't be the first function that doesn't either.

TimothyGu commented on test/parallel/test-util-types.js:
```diff
@@ -0,0 +1,106 @@
+    continue;
+  assert(`is${match[1]}` in types,
+         `util.types should provide check for Is${match[1]}`);
+}
```
Very creative test!

TimothyGu commented on test/parallel/test-util-types.js:
```diff
@@ -0,0 +1,106 @@
+];
+
+if (v8_h === undefined) {
+  process.on('exit', () => common.skip('Could not read v8.h'));
```
`process.on('exit')` wouldn't be necessary if the `WebAssembly.Module` is created synchronously.

vsemozhetbyt commented on doc/api/util.md:
```diff
@@ -1367,15 +1890,42 @@ deprecated: v0.11.3
+[`util.types.isDate()`]: #util_util_types_isdate_value
+[`util.types.isNativeError()`]: #util_util_types_isnativeerror_value
+[`util.types.isSharedArrayBuffer()`]: #util_util_types_issharedarraybuffer_value
+[`TypedArray`]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray
```
Out of ABC-order?

ChALkeR commented on doc/api/deprecations.md:
```diff
@@ -818,6 +818,14 @@ a future version at which point only authentication tag lengths of 128, 120,
+<a id="DEP0XXX"></a>
+### DEP0XXX: process.binding('util').is[...] typechecks
+
+Type: Documentation-only/`--pending-deprecation`
```
Is `--pending-deprecation` the correct type here? It's the first such one, afaik.
/cc @jasnell, I suppose.

Also, `--pending-deprecation` (and it being a `semver-minor`) is not mentioned in [COLLABORATOR_GUIDE.md#deprecations](https://github.com/nodejs/node/blob/master/COLLABORATOR_GUIDE.md#deprecations), but that's probably a separate issue.

ChALkeR replied:
I opened #18417 for that.

addaleax replied:
> As I understand, currently `--pending-deprecation` is just a subclass of doc-only deprecations.

I’m okay with seeing it that way, yeah. And I don’t mind being explicit about this here.

ChALkeR replied:
#18433 landed, so could you change this line to 
```
Type: Documentation-only (supports [`--pending-deprecation`][])
```
for consisntency?

vsemozhetbyt commented on doc/api/util.md:
````diff
@@ -894,6 +894,526 @@ encoded bytes.
+
+```js
+const target = {};
+const proxy = new Proxy({}, {});
````
Maybe `const proxy = new Proxy(target, {});` to make `target` more target?)

vsemozhetbyt commented on doc/api/util.md:
```diff
@@ -1367,15 +1890,42 @@ deprecated: v0.11.3
+[`Uint8ClampedArray`]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8ClampedArray
+[`Uint16Array`]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint16Array
+[`Uint32Array`]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint32Array
+[`WeakSet`]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakSet
```
Should be after `WeakMap`.

vsemozhetbyt commented:
Maybe it is due to big diff, but some old review comments are not hidden now and are associated with the wrong lines, so do not be confused:

https://github.com/nodejs/node/pull/18415#discussion_r164285093 (or https://github.com/nodejs/node/pull/18415/files#r164285093)

https://github.com/nodejs/node/pull/18415#discussion_r164287166 (or https://github.com/nodejs/node/pull/18415/files#r164287166)

jasnell commented on doc/api/util.md:
```diff
@@ -894,6 +894,534 @@ encoded bytes.
 
 The encoding supported by the `TextEncoder` instance. Always set to `'utf-8'`.
 
+## util.types
```
Just noting that this will have a named-import-in-ESM issue just like the `fs.promises` addition.

jasnell commented on lib/internal/util/types.js:
```diff
@@ -29,8 +29,48 @@ function isUint8Array(value) {
 }
 
+function isUint8ClampedArray(value) {
+  return TypedArrayProto_toStringTag(value) === 'Uint8ClampedArray';
```
We should likely note in the documentation how these checks are performed because types can lie about what they are if the user overrides `Symbol.toStringTag`.

TimothyGu replied:
This function is cached at Node.js startup and thus immune to userland overrides.

jasnell replied:
Ah right right... completely zoned on the caching. Just verified locally that even preload script can't lie about this. Dismiss this comment then :-)

jasnell replied:
Actually... @addaleax ... this would be something worth adding to the tests... specifically, ensuring that a user-provided class overriding a built-in set within a preload module cannot lie and that the checks will come up as false.

addaleax replied:
>  a user-provided class overriding a built-in set within a preload module cannot lie 

I’m not sure what you mean by that. I have added a test that checks that overriding `Symbol.toStringTag` doesn’t fool this.

jasnell commented on test/parallel/test-util-types.js:
```diff
@@ -0,0 +1,99 @@
+    if ((types.isArrayBufferView(value) ||
+         types.isAnyArrayBuffer(value)) && key.includes('Array')) {
+      continue;
+    }
```
The test should also likely verify that the result is consistent across vm contexts... e.g. check that `vm.runInNewContext('new Map()')` passes the type check.

jasnell commented on doc/api/util.md:
```diff
@@ -894,6 +894,534 @@ encoded bytes.
+util.types.isBooleanObject(false);  // Returns false
+util.types.isBooleanObject(true);   // Returns false
+util.types.isBooleanObject(new Boolean(false));   // Returns true
+util.types.isBooleanObject(new Boolean(true));    // Returns true
```
add:

```
util.types.isBooleanObject(Boolean(false)); // Returns false
util.types.isBooleanObject(Boolean(true)); // Returns false
```

jasnell commented on doc/api/util.md:
```diff
@@ -894,6 +894,534 @@ encoded bytes.
+added: REPLACEME
+-->
+
+Returns `true` if the value is a built-in [`DataView`][] instance.
```
Not sure if we should assert that it is 'built-in' in these. The objects can lie, and user code can create their own objects (polyfills) that are not actually built-ins.

jasnell commented on doc/api/util.md:
```diff
@@ -894,6 +894,534 @@ encoded bytes.
+added: REPLACEME
+-->
+
+Returns `true` if the value is an async function.
```
`async function` here could be a link to more documentation about async functions... perhaps a link over the MDN documentation?

jasnell commented on doc/api/util.md:
```diff
@@ -894,6 +894,534 @@ encoded bytes.
+
+The result generally does not make any guarantees about what kinds of
+properties or behavior a value exposes in JavaScript. They are primarily
+useful for addon developers who prefer to do type checking in JavaScript.
```
hmm... this last sentence is a bit content free. *why* would developers prefer to do type checking in JavaScript :-) ... I think it's better to say something about specific use cases such as "These are primarily useful for debugging" or something similar

addaleax replied:
@jasnell I can’t really agree … the “why” is, it’s a preference. I personally don’t think it’s a good idea, but it’s what Node decided to migrate to?

Saying “these are primarily useful for debugging” doesn’t really seem more helpful, because it doesn’t even say how you’d use these methods for debugging…

TimothyGu commented on test/parallel/test-util-types.js:
```diff
@@ -0,0 +1,99 @@
+/*global SharedArrayBuffer*/
+'use strict';
```
There's already https://github.com/nodejs/node/blob/master/test/parallel/test-types.js. You might want to look into combining these two files.

addaleax replied:
@TimothyGu I’ve just copied the contents into a block in this file, let me know if that doesn’t sound good to you. The tests have somewhat different structures, but I think that’s fine?

addaleax commented:
@jasnell @TimothyGu @ChALkeR Could you take another look?

TimothyGu commented:
These two comments of mine don't seem to be addressed:

- https://github.com/nodejs/node/pull/18415#discussion_r164285667
- https://github.com/nodejs/node/pull/18415#discussion_r164306278

addaleax commented:
@TimothyGu You’re right. In the `GeneratorObject` case I wasn’t sure whether your comment would be blocking, but then again if V8 exposes that API I guess there’s no harm including it here.

ChALkeR commented on src/node_types.cc:
```diff
@@ -0,0 +1,69 @@
+
+#define V(type) \
+  static void Is##type(const FunctionCallbackInfo<Value>& args) {             \
+    args.GetReturnValue().Set(args[0]->Is##type());                           \
```
If this gets exported and documented, there should be proper checks for the size of `args`, I suppose.

What would `util.types.isMap()` do without arguments?

TimothyGu replied:
It would treat the argument as `undefined`. I’m not sure if this is the intended behavior.

ChALkeR replied:
@TimothyGu Ah, that's good, then.

BridgeAR commented:
Ping @addaleax this needs a rebase but I think otherwise this is ready?

addaleax commented:
Rebased!

CI: https://ci.nodejs.org/job/node-test-commit/16659/

addaleax commented:
Landed in bd6e0be0dfbbf6d10f1f054c43b0b82f25c60b16, b20af8088a4d5cccb1901d42107f6d9ca82d306f, fixed the linter failure from the rebase while landing.

MylesBorins commented:
Should this be backported to `v9.x-staging`? If yes please follow [the guide](https://github.com/nodejs/node/blob/master/doc/guides/backporting-to-release-lines.md) and raise a backport PR, if not let me know or add the `dont-land-on` label.

MylesBorins commented:
this would need to come with https://github.com/nodejs/node/pull/19149

MylesBorins commented:
Is this something we would want to backport to v8.x? Seems like a bunch of moving parts
